scala/scala-impl/src/org/jetbrains/plugins/scala/lang/psi/impl/expr/ScForImpl.scala

```diff
@@ -183,8 +183,15 @@ class ScForImpl(node: ASTNode) extends ScExpressionImplBase(node) with ScFor wit
       normalizeUnderscores(expr).getText
     }
 
-    def needsPatternMatchFilter(pattern: ScPattern): Boolean =
-      !pattern.isIrrefutableFor(if (forDisplay) pattern.expectedType else None)
+    def needsPatternMatchFilter(pattern: ScPattern): Boolean = {
+      val hasSc3Case = pattern.isScala3OrSource3Enabled &&
```

Comment:
It seems like the check **is** required, but currently not covered with tests.
It will be more clear once [comment](https://github.com/JetBrains/intellij-scala/pull/660/files#r1584668155) is addressed